scipy/spatial/transform/rotation.py

```diff
@@ -3,12 +3,127 @@
+    # Algorithm assumes axes as column vectors, here we use 1D vectors
+    n1 = elementary_basis_vector(seq[0])
+    n2 = elementary_basis_vector(seq[1])
+    n3 = elementary_basis_vector(seq[2])
+
+    # Step 2
+    sl = np.dot(np.cross(n1, n2), n3)
+    cl = np.dot(n1, n3)
+
+    # angle offset is lambda from the paper referenced in [2] from docstring of
+    # `as_euler` function
+    offset = np.arctan2(sl, cl)
+    c = np.empty((3, 3))
+    c[0] = n2
+    c[1] = np.cross(n1, n2)
+    c[2] = n1
+
+    # Step 3
+    rt = np.array([
+        [1, 0, 0],
+        [0, cl, -sl],
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
+    res = np.einsum('...ij,...jk->...ik', rtc, dcm)
+    dcm_transformed = np.einsum('...ij,...jk->...ik', res, c.T)
+
+    # Step 4
+    angles = np.empty((num_rotations, 3))
+    angles[:, 1] = np.arccos(dcm_transformed[:, 2, 2])
+
+    # Steps 5, 6
+    eps = 1e-7
+    safe1 = (np.abs(angles[:, 1]) >= eps)
+    safe2 = (np.abs(angles[:, 1] - np.pi) >= eps)
+
+    # Step 4 (Completion)
+    angles[:, 1] += offset
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angles[safe_mask, 0] = np.arctan2(dcm_transformed[safe_mask, 2, 0],
+                                      -dcm_transformed[safe_mask, 2, 1])
+    angles[safe_mask, 2] = np.arctan2(dcm_transformed[safe_mask, 0, 2],
+                                      dcm_transformed[safe_mask, 1, 2])
+
+    # 6a
+    angles[~safe_mask, 2] = 0
+    # 6b
+    angles[~safe1, 0] = np.arctan2(
+        dcm_transformed[~safe1, 0, 1] - dcm_transformed[~safe1, 1, 0],
+        dcm_transformed[~safe1, 0, 0] + dcm_transformed[~safe1, 1, 1]
+    )
+    # 6c
+    angles[~safe2, 0] = np.arctan2(
+        dcm_transformed[~safe2, 0, 1] + dcm_transformed[~safe2, 1, 0],
+        dcm_transformed[~safe2, 0, 0] - dcm_transformed[~safe2, 1, 1]
+    )
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
+    if seq[0] == seq[2]:
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angles[:, 1] < 0, angles[:, 1] > np.pi)
+    else:
+        # lambda = + or - pi/2, so we can ensure angle2 -> [-pi/2, pi/2]
+        adjust_mask = np.logical_or(angles[:, 1] < -np.pi / 2,
+                                    angles[:, 1] > np.pi / 2)
+
+    # Dont adjust gimbal locked angle sequences
+    adjust_mask = np.logical_and(adjust_mask, safe_mask)
+
+    angles[adjust_mask, 0] += np.pi
+    angles[adjust_mask, 1] = 2 * offset - angles[adjust_mask, 1]
+    angles[adjust_mask, 2] -= np.pi
+
+    angles[angles < -np.pi] += 2 * np.pi
+    angles[angles > np.pi] -= 2 * np.pi
+
+    # Step 8
+    # TODO: if any observability flags are poor, possibly raise a UserWarning?
+    if not np.all(safe_mask):
+        warnings.warn("Some angle sequences suffer from gimbal lock. In those "
+                      "cases, it is not possible to determine the angles "
+                      "uniquely.", UserWarning, stacklevel=1)
```

Comment:
Perhaps add "the third angle was set to 0". Also what about those setting `UserWarning, stackleve=1`? Why you decided to use these?

If possible, could you try to shorten the warning message? I think shorter the message --- easier for a user to quickly read it and understand.

Reply:
`UserWarning` is the default warning class. I decided to state that because I'm checking for that particular class in the tests.
Regarding the `stacklevel=1`, I thought it would be appropriate if the warning was raised at the `as_euler` function. Maybe I'm wrong here.

I'll remove the stack level and add the message about the third angle.